staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go

```diff
@@ -0,0 +1,211 @@
+package csaupgrade
+
+import (
+	"bytes"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+)
+
+const csaAnnotationName = "kubectl.kubernetes.io/last-applied-configuration"
+
+var csaAnnotationFieldSet = fieldpath.NewSet(fieldpath.MakePathOrDie("metadata", "annotations", csaAnnotationName))
+
+// Upgrades the Manager information for fields managed with client-side-apply (CSA)
+// Prepares fields owned by `csaManager` for 'Update' operations for use now
+// with the given `ssaManager` for `Apply` operations.
+//
+// This transformation should be performed on an object if it has been previously
+// managed using client-side-apply to prepare it for future use with
+// server-side-apply.
+//
+// Caveats:
+//  1. This operation is not reversible. Information about which fields the client
+//     owned will be lost in this operation.
+//  2. Supports being performed either before or after initial server-side apply.
+//  3. Client-side apply tends to own more fields (including fields that are defaulted),
+//     this will possibly remove this defaults, they will be re-defaulted, that's fine.
+//  4. Care must be taken to not overwrite the managed fields on the server if they
+//     have changed before sending a patch.
+//
+// obj - Target of the operation which has been managed with CSA in the past
+// csaManagerName - Name of FieldManager formerly used for `Update` operations
+// ssaManagerName - Name of FieldManager formerly used for `Apply` operations
+// Returns: a copy of the `obj` paramter with its managed fields modified so that
```

Comment:
Any reason not to do it in-place? People can decide to deepcopy the object beforehand if they want to keep a copy of it.

Reply:
changed to in-place